### PR TITLE
revises turbine epoch stakes for shreds propagation

### PR DESCRIPTION
**Pull request by behzadnouri**
#### Problem
`Bank::get_leader_schedule_epoch` returns: `1 + EpochSchedule::get_epoch`. As a result, at the epoch boundaries when the propagated shred is from the next epoch, we are looking for epoch stakes for 2 epochs ahead of the root or working bank's epoch.
However, the bank structure only contains epoch stakes for one epoch ahead. This results in shreds propagated at epoch boundary having unknown epoch stakes.

#### Summary of Changes
Use `EpochSchedule::get_epoch` instead of  `Bank::get_leader_schedule_epoch` for turbine epoch stakes when propagating shreds.

Previous discussion on this: https://github.com/solana-labs/solana/pull/18971#discussion_r681930509

**NOTE:**
The commit effectively uses current epoch instead of next epoch for stakes.
One question is that if that may cause more dead stake which would cause holes in turbine propagation tree?
e.g. if some nodes destake, then they will be removed from the next epoch stakes but not the current one.
In other words, how many epochs it will take for a destaked node not to be used in turbine propagation anymore?

### CODE REVIEW

File: turbine/src/cluster_nodes.rs [hunk 1]
```diff
@@ -362,7 +362,15 @@ impl<T: 'static> ClusterNodesCache<T> {
         working_bank: &Bank,
         cluster_info: &ClusterInfo,
     ) -> Arc<ClusterNodes<T>> {
-        let epoch = root_bank.get_leader_schedule_epoch(shred_slot);
+        let epoch = if check_feature_activation(
+            &feature_set::revise_turbine_epoch_stakes::id(),
+            shred_slot,
+            root_bank,
```
Comment (t-nelson) [inline turbine/src/cluster_nodes.rs:368]: seems like it's ok in this case that the feature activation is observed after the actual epoch boundary, since schedule and activation status are queried from the same bank?

**Review by t-nelson — DISMISSED** [review]
nice catch!

> One question is that if that may cause more dead stake which would cause holes in turbine propagation tree?
e.g. if some nodes destake, then they will be removed from the next epoch stakes but not the current one.
In other words, how many epochs it will take for a destaked node not to be used in turbine propagation anymore?

outside some abnormal market events, i suspect that stake deactivation vs deactivation will effectively balance out. we'll always lag a bit, but at least using the current epoch stakes, there will be less ability to manipulate the distribution with transient stake cancellation. the worse case is staked nodes shutting down without destaking, but that's unpredictable

**Comment by behzadnouri** [comment]
> but at least using the current epoch stakes, there will be less ability to manipulate the distribution with transient stake cancellation.

How would transient stake cancellation impact epoch-stakes?

My understanding is that epoch-stakes are computed and immediately frozen at the epoch boundary for the _next_ epoch:
https://github.com/solana-labs/solana/blob/b44c9bca8/runtime/src/bank.rs#L1536-L1541
https://github.com/solana-labs/solana/blob/b44c9bca8/runtime/src/bank.rs#L2187-L2219
so stake changes _within_ the epoch shouldn't impact epoch-stakes unless they persist until the epoch boundary.

File: turbine/src/cluster_nodes.rs (hunk 1 above)
Comment (t-nelson) [reply turbine/src/cluster_nodes.rs:368]: nope. https://github.com/solana-labs/solana/blob/b44c9bca898122bd3b6de24770c2a0acd31da222/runtime/src/bank.rs#L1539

**Comment by t-nelson** [comment]
> My understanding is that epoch-stakes are computed and immediately frozen at the epoch boundary for the _next_ epoch: https://github.com/solana-labs/solana/blob/b44c9bca8/runtime/src/bank.rs#L1536-L1541 https://github.com/solana-labs/solana/blob/b44c9bca8/runtime/src/bank.rs#L2187-L2219 so stake changes _within_ the epoch shouldn't impact epoch-stakes unless they persist until the epoch boundary.

ah right, we project the current weights one ahead and latch it in. n/m that comment. :+1:

File: turbine/src/cluster_nodes.rs (hunk 1 above)
Comment (behzadnouri) [reply turbine/src/cluster_nodes.rs:368]: that is not epoch-schedule though.
did you mean epoch-stakes in your first comment?

Comment (t-nelson) [reply turbine/src/cluster_nodes.rs:368]: ah right.  should be fine so long as we don't happen to be re-parameterizing the epoch schedule via feature gate at the same time

File: turbine/src/cluster_nodes.rs [hunk 2]
```diff
@@ -362,7 +362,15 @@ impl<T: 'static> ClusterNodesCache<T> {
...
+        ) {
+            root_bank.epoch_schedule().get_epoch(shred_slot)
```
Comment (t-nelson) [inline turbine/src/cluster_nodes.rs:370]: do we need to also use `epoch_schedule()` below in the fallback case where epoch-stakes aren't available from either `root_bank` or `working_bank` for the shred slot and we use epoch-stakes from the root_banks's slot instead?

Comment (behzadnouri) [reply turbine/src/cluster_nodes.rs]: done.

**Review by t-nelson — DISMISSED** [review]
thanks!

:shipit:

**Review by carllin — APPROVED** [review]

**Comment by behzadnouri** [comment]
Can I have another :stamp: here?
I had to rebase due to merge conflicts.

**Review by t-nelson — APPROVED** [review]
